Replace NNI with auto-sklearn, and add some testing

benclifford opened the pull request:
This PR:
* uses auto-sklearn instead of NNI (although it does not remove the NNI code)
* adds a testing script
* makes the code run from the root of a blindml checkout, rather than /Users/maksim

benclifford committed:
Replace hardcoded ~maksim paths with .

This probably means that the code will only work run from the blindml source directory.

but it's better than not running.

benclifford committed:
Add auto-sklearn requirement

benclifford committed:
pin nni=1.9 because more recent versions break some package layout at least

tooling as installed by this is broken right now for me in a clean venv
- pip-compile is broken because pip-tools breaks with recent pip, and apparently
the requirements.txt doesnt pin that stuff hard enough.

- the breakage only happens when I try to add in auto-sktools to requirements.txt
and rebuild...

this is an upgrade of pip-tools and a rerun of the regenerate/reinstall until
a fixpoint is reached in my env

first manually edit requirements.txt because pip-compile cant do this itself to
constrain pip-tools>=5.5

then:
run this over and over until no more git diff:

(looks like one iteration is ok for me actually)

rm -rfv ~/cqx/c/datastation/virtualenv/ && virtualenv /home/benc/cqx/c/datastation/virtualenv/ --python=python3 && pip install -r requirements.txt --extra-index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple && pip-compile --extra-index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple

benclifford committed:
Replace NNI calls with auto-sklearn calls

benclifford committed:
Add flake8 dep and config, with disables to disable what is already broken

This gives basic linting for style errors and some basic static semantic checking.

Future commits could tidy up the broken flake8 tests and re-enable them.

Run with:

  $ flake8

benclifford committed:
Add statistical test for basic sanity of behaviour

Remove existing test that fails for me

Adjust path of other existing test to run from blindml repo root for when it is invoked by pytest there

benclifford committed:
fix flake8

makslevental commented on blindml/frontend/config/task/task.py:
```diff
@@ -75,6 +77,8 @@ def __init__(self, task_fp) -> None:
                     - set(self._task.payload.drop_cols)
                     - {self._task.payload.y_col}
                 )
+                # ^ should this drop the y_col even when there isn't a drop_cols
+                # configuration?
```
yes because otherwise we're training on the `y_col` to predict `y_col` (i.e. poisoning the training)

benclifford replied:
i guess if the user specifies the ycol in their explicitly specified X_cols, that's their own fault, though?

makslevental replied:
yup major party foul (but there should also probably be a check)

benclifford replied:
it looks like the explicit X_cols code path was broken anyway - I've put in a fix and added a test case.

I've also added in an assert in the explicit X_cols path for "no ycol in xcols"

makslevental commented on blindml/frontend/config/task/task.py:
```diff
@@ -101,38 +105,35 @@ def __init__(self, task_fp) -> None:
-
-    def train_best_model(self):
-        model = self.get_best_model()
+        # this only does regressions. choice of model is defined in
+        # the task defininition, though
+        # TODO: pick model type from there from there
```
can you please include in this PR (i.e. support classification as well)

benclifford replied:
I've added in classification support along with a very simple (6 row) test dataset in pytest.

I couldn't (in the short time I spent) get this to work with NNI, so behaviour may be different from before.

makslevental commented on blindml/frontend/config/task/task.py:
```diff
@@ -101,38 +105,35 @@ def __init__(self, task_fp) -> None:
+
+        # TODO: should give this both training and test data?
+        # which Task has already split up
+        # or give it all data and let regressor split it all
+        # up itself?
```
[their example indicates that autosklearn does not split up the data](https://automl.github.io/auto-sklearn/master/#example)

```
    X_train, X_test, y_train, y_test = \
            sklearn.model_selection.train_test_split(X, y, random_state=1)
    automl = autosklearn.classification.AutoSklearnClassifier()
    automl.fit(X_train, y_train)
```

so i think the answer is no here

makslevental commented on blindml/frontend/config/task/task.py:
```diff
@@ -101,38 +105,35 @@ def __init__(self, task_fp) -> None:
+        print(regressor.show_models())
+
+        self._auto_sk_model = regressor
+        # this will resume? if experiment already exists?
+        # run_nni(self._nni_experiment_config)
```
please remove

makslevental commented on blindml/runner.py:
```diff
@@ -5,24 +5,16 @@
-    while not task.get_model_search_update():
-        print("no model trained yet")
-        time.sleep(5)
+    model = task_auto_sk_model
```
`task_auto_sk_model` is not defined in this function (nor in this file...)

benclifford replied:
missing . fixed in 7c840fa9ee580e883d7d30a5758ac415c24fc8be

makslevental commented on requirements.in:
```diff
@@ -1,10 +1,11 @@
 keras
 tensorflow
-nni
+nni==1.9
```
please remove

makslevental commented:
Should've put this in the review whoops: please remove references to `nni` (including helper/util code)

benclifford committed:
Add classification support to PR #4

Prior to this, auto-sklearn was only called in regression mode

can i get a classification test/demo?

does the previous NNI implementation in blindml support classification, to test my changes against?
I couldn't get it to work on a simple csv that i made up

benclifford committed:
Remove TODO/question/comment that is answered

benclifford committed:
remove some commented out code

benclifford committed:
add missing dot

benclifford committed:
Remove nni prereq package, and everything that breaks in quick-test.sh because of that

benclifford committed:
Fix broken explicit-X_cols codepath and add a test

benclifford committed:
check that y col is not in x cols when explicitly specified

benclifford committed:
fix flake8 for quick test
